Add foreigntypes package to hold FFI interop types

dipinhora opened the pull request:
NOTE: This PR has been updated to only include the `foreigntypes` package. See discussion thread on this PR for details.

-----------------

The motivation behind this is to improve performance to file i/o
operations. The C standard stdio functions do user space
buffering prior to sending the data to the OS kernel via a system
call. On most operating system, the OS kernel also does additional
buffering on top of the user level buffering by the C functions.
Unfortunately, if the C functions are buffering data, they prevent
the OS kernal from taking advantage of some potential performance
enhancements if they're available (namely Vectored I/O so the
OS can offload work to hardward and never has to combine multiple
user buffers into a single buffer for writing to disk).

The commit changes the File class to use low level i/o functions
that map to system calls on Linux and other posix based systems
and to C lib calls on Windows. On Posix based systems, the OS is
free to take advantage of performance offloading to hardware if
possible. On Windows, this change shouldn't have a negative
performance impact but doesn't give the same possible performance
benefits as on Posix based systems.

Other changes:

* Enable `_FILE_OFFSET_BITS=64` arg to compiler to ensure 32 bit
  linux builds get large file support enabled.
* Adds `datasync` to be able to call `fdatasync` instead of `fsync`
  as it can be appropriate in some situations. On Windows, `fsync`
  and `fdatasync` behave the same.
* Adds `queue` and `queuev` functions that are analogous to
  `write` and `writev` but queue/buffer data to be written
  instead of writing the data immediately. Additionally,
  these new functions provide the ability to interleave buffered
  data writing and unbuffered data writing where unbuffered data
  writing would force any buffered data to be written first.
* Changes `flush` to flush any queued/buffered data to disk
  because it is otherwise unnecessary since we're not buffering
  data in the C library via stdio any longer.
* A few additional tests have been added to validate things are
  working correctly.

Future work:
* Change `queue`/`queuev` to automatically flush data to disk once
  the # of queued items exceeds `IOV_MAX` for a single system call.
* Error handling in the File package might be possible to add now
  including having the various functions throw `error` on things
  that can't be recovered from.
* Possibly switch to async/nonblocking i/o for file writing with an
  actor as the intermediary (similar to the TCPConnection actor in
  the net package).

--------

Also added foreigntypes package to hold FFI interop types

The new package is to help with consistency of data types
mapping between Pony and FFI types. Initially includes types
related to the Posix standard.

Thanks to @sylvanc for help and feedback regarding this.

--------

The following is a great explanation of why `writev` is worth it (from https://bytes.com/topic/python/answers/39551-writev) in reference to python:
```
writelines applies to a general Python file object. writev applies only
to C file descriptors. Writev can't replace writelines, after all it
makes no sense to cStringIO, gzip files for these are not valid C file
descriptors. Generally, writelines works fine.

Now why would you want to use writev? Optimization on the C side.

Understand that when you do file I/O you have to either:

a) Copy all of the strings to a new memory location
b) Call write over and over again

Sometimes, when you do b), userspace libraries (fwrite) will "optimize"
by buffering and doing a) for you. But you cannot escape the fact that
so long as your write parameter takes a single string for each
invocation, the underlying libraries are forced to choose between the
two options above.

Writev is the vector version of write. Whereas write accepts a single
pointer and length parameter, writev accepts a *list* of pointers and
size parameters. It represents a strategy c) -- just hand the list to
operating system

I want to include it because POSIX has a single OS call that
conceptually maps pretty closely to writelines. writev can be faster
because you don't have to do memory copies to buffer data in one place
for it -- the OS will do that, and can sometimes delegate that chore to
the underlying network or scsi card.

If you are still scratching your head, just think of writev as a C-file
descriptor only optimization of writelines that offloads the memory copy
cost of buffering to the OS in hopes that it can pass the buck to the
hardware (and IIRC, BSD does handle this correctly ... it has zero copy
network code, think about how cool it is to think that your network card
is going to traverse your list for you with a little help from the
writev function.)
```

dipinhora commented:
Forgot to mention:

It would be ideal to be able to call `dispose` from `_final` in the `File` class so that any open files can be flushed/synced/closed correctly but currently that is not possible because `_final` can't call any `ref` functions.

jemc commented on packages/foreigntypes/posixtypes.pony:
```diff
@@ -0,0 +1,95 @@
+
+// not consistent across OSX (always 32 bit int)/Linux (Long)
+type BlkSize is ILong
+type OSXBlkSize is I32
```
I'm not a big fan of this way of distinguishing between the Linux size and the OSX size.

I'd rather see something like this:
```pony
ifdef linux then
  type BlkSize is ILong
elseifdef osx then
  type BlkSize is I32
end
```

However, that is not currently valid Pony syntax.  I think we should take a look at whether such syntax could be supported, and pursue that alternative.

I suppose this PR could still be merged as-is with this implementation as a stop-gap until we get those ifdefs, but I'll be grumbly about it :disappointed:.

sylvanc replied:
This was my crappy suggestion. The problem is that ponyc tries very hard indeed to type check on all platforms, even though it compiles for just one at a time. @andymcn put a ton of work into that. Adding platform specific types would be... tricky. I'm all for it in theory, but it seems to constitute a combinatorial attack on the type checker :)

dipinhora replied:
@jemc What you suggested is how I initially tried it and the compiler complained.

I think we're likely going to have to merge as is until Pony's type checker is enhanced as @sylvanc mentioned.

jemc replied:
@sylvanc - if we can eliminate dead `ifdef` branches eagerly (before the `names` pass), it doesn't seem like this would add any additional work for the type checker, and would be quite simple to handle, but maybe I'm missing something.

The main barrier to this that I'm seeing is that it may be difficult to integrate this strategy with the parser.  Because the Pony parser is LL1, and type entity declarations and method bodies are parsed without a ending token, Pony relies on seeing the next `fun`, `new`, `class`, `actor`, `primitive`, `type`, or `struct` token to know that current method body is finished (or type entity declaration) is finished.  If we wanted to allow `ifdef`s at the top level scope of a file governing the selective compilation of types (or for that matter, within a type entity declaration governing the selective compilation of methods), we'd have to figure out something clever with the parser and/or the `syntax` pass that follows the parser to make it work.

One slightly ugly, but quite easy option would be to introduce a new keyword, (for example, `typeifdef`) for `ifdef`s that need to be used at the top level scope.  I don't really like this option, but I wanted to mention it.

Another option would be to let an `if` follow on the same line as the type, just as we do for `use` calls.
```pony
type BlkSize is ILong if linux
type BlkSize is I32 if osx
```

jemc commented on packages/foreigntypes/posixtypes.pony:
```diff
@@ -0,0 +1,95 @@
+type OSXClock is ULong
+
+// not defined on OSX
+type ClockId is I32
```
Again, this is a case where I'd rather be able to use `ifdef`s to enforce that it isn't defined on OSX.

sylvanc replied:
...which we can't do at the type level, currently.

jemc commented on packages/foreigntypes/posixtypes.pony:
```diff
@@ -0,0 +1,95 @@
+
+type Id is U32
+
+type INo is U64
```
What C type does this correspond to?  I couldn't find an `i_no_t` in a google search.

In general it would be nice if all of these listed the C named type they correspond to.  Perhaps in the docstring?

```pony
type INo is U64 "i_no_t"
```

If not the docstring, at least a comment would be nice.

dipinhora replied:
It's `ino_t` from the `stat` structure and represents "File serial number."

I didn't know docstrings were possible on type aliases like that. I'll add them.

dipinhora commented on src/libponyrt/lang/io.c:
```diff
@@ -0,0 +1,21 @@
+
+PONY_EXTERN_C_BEGIN
+
+#ifndef PLATFORM_IS_WINDOWS
```
That's way too logical....... I'll make the change. 8*)

dipinhora commented on packages/foreigntypes/posixtypes.pony:
```diff
@@ -0,0 +1,95 @@
+// not consistent across OSX (always unsigned long)/Linux (always unsigned int)
+type Speed is ULong
+
+type SSize is ISize
```
I'm more than happy to remove `SSize` and `Size` also if that's the decision.

dipinhora commented:
@jemc `io.c` fixed and docstrings added.

Let me know what the decision is on `SSize` and `Size` and I'll make the appropriate changes if required.

Also, @jemc and @sylvanc, any thoughts on:

> It would be ideal to be able to call `dispose` from `_final` in the `File` class so that any open files can be flushed/synced/closed correctly but currently that is not possible because `_final` can't call any `ref` functions.

If we could call `dispose` from `_final` we could enable this `Rust`-like pattern (and it would be safer in general):

```
// do things before file is opened
{
    let file = ...;
    // do things with file
}
// file has been closed at this point
```

sylvanc commented:
Finalisation is not deterministic. I think what you want is:

```pony
// do things before file is opened
with file = ... do
  // do things with file
end
// file has been closed at this point
```

The `with` construct disposes of things when the scope ends. You can have more than one thing in a `with`:

```pony
// do things before file is opened
with file1 = ..., file2 = ... do
  // do things with file
end
// file has been closed at this point
```

dipinhora commented:
@sylvanc Using `with` makes sense. I suppose I should get more familiar with actually writing Pony programs as opposed to playing around in the runtime.

In regards to running `dispose` from `_final`, is that something that makes sense? It seems to me from a safety and principle of least surprise perspective it would make sense to ensure that a program has successfully flushed/synced data and closed a file before exiting. But I tend to overthink the edge cases so I could be off base.

sylvanc commented:
`dispose` modified the object. `_final` _must not_ modify the object, or indeed any object that existed prior to finalisation, otherwise finalisation can result in object resurrection. Non-resurrecting finalisation is a big feature for Pony :)

This is why `_final` tends to do the same logical work as `dispose`, but without doing things like setting fields.

sylvanc commented:
@jemc left a comment on the code that showed up in my email, but isn't here. Perhaps since it was a line comment instead of a conversation comment, it was destroyed by a push? In any case, here it is:

> @sylvanc - if we can eliminate dead ifdef branches eagerly (before the names pass), it doesn't seem like this would add any additional work for the type checker, and would be quite simple to handle, but maybe I'm missing something.
> The main barrier to this that I'm seeing is that it may be difficult to integrate this strategy with the parser. Because the Pony parser is LL1, and type entity declarations and method bodies are parsed without a ending token, Pony relies on seeing the next fun, new, class, actor, primitive, type, or struct token to know that current method body is finished (or type entity declaration) is finished. If we wanted to allow ifdefs at the top level scope of a file governing the selective compilation of types (or for that matter, within a type entity declaration governing the selective compilation of methods), we'd have to figure out something clever with the parser and/or the syntax pass that follows the parser to make it work.
> One slightly ugly, but quite easy option would be to introduce a new keyword, (for example, typeifdef) for ifdefs that need to be used at the top level scope. I don't really like this option, but I wanted to mention it.
> Another option would be to let an if follow on the same line as the type, just as we do for use calls.
> type BlkSize is ILong if linux
> type BlkSize is I32 if osx

This would indeed work! But it would run into the problem I was talking about before: we could no longer type check for all platforms.

Right now, it's possible to write cross-platform Pony code, and be sure it typechecks on all platforms. We could give up on that, or possibly come up with a fancy way to keep it in the presence of platform-specific types that didn't lead to slow compile times.

If we were to give up on it, the `if` syntax above is very nice. But I think giving up on it would be unfortunate (although perhaps necessary).

dipinhora commented:
@sylvanc That makes sense re: `dispose`/`_final` (not that I actually understand it fully but zombie objects don't seem like a good thing 8*P).

I'll see if I can figure out how to make `_final` do everything that `dispose` does without duplicating all writing/flushing/syncing/closing logic.

jemc commented:
@sylvanc - I'd argue that having hacks that require you to remember to use `BlkSize` vs `OSXBlkSize` are brittle, hard to extend to include more platforms, and make the program prone to bugs that are difficult and hairy to track down.  We do get some value from knowing that the code all type-checks correctly on all platforms, but as we can see from the variance in these types across platforms, just getting the type checking right at the Pony level is not all there is to correctness here, and I'd argue that the `BlkSize` vs `OSXBlkSize` hacks are directly threatening to correctness, since they are difficult to use.

Confirming that your code type checks on all platforms seems most like a job for package writers, a task that belongs either in CI, or as part of some prior-to-releasing ritual if not using CI.  That is, in your everyday compiling of code for production or testing, you mainly care that it typechecks on *your* platform, and you don't really care about other platforms.  So, in light of that observation, I'd propose the following workaround:

We could add a compiler flag to override platform detection, and simulate a specific platform that you chose to give in the flag.  This would let you confirm that the program typechecks/compiles on that platform.  Package writers could use this from their CI tasks (or prior-to-releasing ritual).

One possible caveat is that users might think they were cross-compiling correct machine code by using that flag, so maybe we would only want to allow using that flag if the pass limiter was also enabled?

sylvanc commented:
>  I'd argue that having hacks that require you to remember to use BlkSize vs OSXBlkSize are brittle, hard to extend to include more platforms, and make the program prone to bugs that are difficult and hairy to track down.

And I argue that... you are right! Completely and utterly. But cross-platform type checking isn't in opposition to this - it's actually the same exact goal.

It's true that type checking is not formal verification (i.e. there is more to correctness than type checking). But giving up on type checking doesn't help. Your point that this could be handled by re-running the typechecker for every possible configuration is exactly the combinatorial attack on the type system that I'm worried about :) I may indeed have to give up on this worry, but I would much prefer to have a system that took advantage of the existing work on cross-platform type checking.

The specific example of `blksize_t` is interesting, because it is only used in `struct stat`, which is a very platform-specific structure. So platform specific that it really needs the C-header via libclang RFC in order to be handled in Pony. At which point we would have something sort of vaguely like this:

```pony
use "sys/stat.h" // all platforms
use "sys/types.h" if not osx // not on MacOS
use "unistd.h" if linux // only on linux

struct Stat \type=stat if not windows, type=_stat if windows\
  // problem: no such field on Windows
  // other problem: might need platform-specific type for MacOS
  // other other problem: this is filthy
  let block_size: BlkSize \field=st_blksize\
```

The horror above isn't a Pony issue, but a C issue. Importing C-header `#ifdef` hell makes me nervous.

I'm not offering a solution, btw, just wondering if anybody else can :)

jemc commented:
> Your point that this could be handled by re-running the typechecker for every possible configuration is exactly the combinatorial attack on the type system that I'm worried about

Right, but with my proposed workaround it would not affect normal compile times - it would only affect the overall time required for a package author to verify compilation on all supported platforms for the package.

I agree it's not an ideal solution, and I'm also open to hearing other ideas.

Honestly, I'd prefer not to merge this PR with the `foreigntypes` package as-is, until we at least have a consensus on a plan for how we want to fix the ugliness in the future.

dipinhora commented:
@sylvanc I have a refactored version of the `File` class where the `_final` can flush data via a shared `fun` (that doesn't write to any fields) that `_pending_writes` also uses to avoid duplicating the logic.

But I am getting the following error now:

```
Error:
/home/vagrant/dhponyc/packages/files/file.pony:700:3: _final cannot create actors or send messages
  fun _final() =>
  ^
    Info:
    /home/vagrant/dhponyc/packages/files/file.pony:710:25: a message can be sent here
              _write_to_disk()
                            ^
    /home/vagrant/dhponyc/packages/files/file.pony:715:21: a message can be sent here
            if path.caps(FileSync) then
                        ^
```

Any pointers on to how to resolve the issues?

sylvanc commented:
Unfortunately, the code that checks `_final` methods to make sure no messages are sent is quite conservative. Two options:

1. Have a look at `finalisers.c`, particularly `check_body_send` and the functions it calls, and see if it can be made more precise (and so allow more things). Calls through traits/interfaces are tricky because methods don't make any guarantees to the type system about whether or not they may send messages.
2. Call cleanup stuff directly from `_final`, avoiding calls through other objects. This is how the existing finalisers work, in `net/ssl` and `regex`. Not ideal, but works.

dipinhora commented:
Thanks @sylvanc. I went for the easy route and fixed the `_final` code to not trigger the `can send message` check (effectively couldn't use `Range` or `Flag` from `collections`). I'm fairly not annoyed with the `_final` implementation now.

@jemc @sylvanc I understand there's still the open question in regards to the `foreigntypes` package and the differences in types for different operating systems. Can we merge this PR as is and figure that out later (I can create an issue to track it)? If not, should I rework this PR to remove `foreigntypes` for now so the `File` class changes can be merged and create a separate PR or issue for `foreigntypes` where discussion can continue?

jemc commented:
I don't know how others feel, but to be honest, I personally am not very comfortable with the way the `foreigntypes` package currently is, and the prospect of merging it as is.

If it could be separated from the other changes, that would probably be a good idea if you wanted to merge the `File` changes soon.

dipinhora commented:
@jemc I split the `File` class changes into PR #1598 (and renamed this PR to be focused around the `foreigntypes` package) because it seems that the `foreigntypes` package will require additional discussion followed by some compiler changes. It doesn't make sense to hold up the `File` changes for an indeterminate amount of time since they are mostly independent of the `foreigntypes` package.

@jemc @sylvanc Let me know how you want to proceed with the `foreigntypes` package. Should we leave open this PR for tracking? Should it be turned into an `Issue` or an `rfc` instead?

jemc commented:
I've been thinking a bit more about this, and I've thought of an alternative proposal that could resolve some of my biggest concerns about the current approach, and not require any new features in the compiler, such as support for `ifdef` clauses governing type definitions.

I'd suggest something like the following approach:

```pony
use linux = "foreigntypes/linux"
use osx = "foreigntypes/osx"
```
```pony
// later in the same file
ifdef linux then
  pseudocode_for_something_linux_specific(linux.BlkSize(0))
elseifdef osx then
  pseudocode_for_something_osx_specific(osx.BlkSize(0))
else
  // ...
end
```

To me, this works much cleaner, because the same `osx` and `linux` identifiers used in `ifdef`s can be used as a package reference, making it quite clear and clean to use those types from within `ifdef`s.

The downside is that there will be some duplication between the sub-packages for different platforms, but I think it's a small price to pay for having the naming convention be so much cleaner, and so much more extensible.

Ideally, we'd have a `foreigntypes` subpackage for every platform recognized by `ifdef`, even for platform specifications like `posix` which represent a grouping of multiple specific platforms (including `osx`, `linux`, `bsd`, etc).  We could potentially define all of the types that are the same across all posix platforms in the `foreigntypes/posix` package, and then the `osx` and `linux` packages could simply alias and reference them so that they would be available from there as well:

```pony
// in foreigntypes/linux package
use posix = "../posix"

type Pid is posix.Pid
```

```pony
// in foreigntypes/osx package
use posix = "../posix"

type Pid is posix.Pid
```

Again, this does add significant extra lines of code, but being able to use a package that was designed to correlate exactly from the `ifdef` you are guarding with is extremely valuable from a code-cleanliness standpoint, in my opinion.

@dipinhora, @sylvanc, @SeanTAllen, @Praetonus, any thoughts on this proposal?

dipinhora commented:
@jemc I like your idea and agree that it is good to have packages per platform. Both for the extra guard/safety of the `ifdef` and because there will invariably be platform specific types that'll need to be defined at some point.

However, I think you have the relationship between the platform specific types and the `posix` general types backwards. The `posix` type for `Pid` should be the same as the platform type that the program is built for since the Posix standard doesn't define the specific types required but just minimum requirements the type must support and lets each platform choose whichever type that they want to use.

It'll need to be something like this instead:

```Pony
// in foreigntypes/posix package
use linux = "../linux"
use osx = "../osx"

// Not legal Pony syntax
type Pid is linux.Pid if linux
type Pid is osx.Pid if osx

// -- OR --

// Not legal Pony syntax
type Pid is ifdef linux then
              linux.Pid
            elseifdef osx then
              osx.Pid
            end
```

This brings us back to the original issue of needing compiler changes to do the `foreigntypes` package properly.

jemc commented:
@dipinhora I was suggesting that types which have different definitions on different Posix platforms not be included in the `posix` package, and that it would only include types which had the same definition on all Posix-based platforms that pony supports.

I understand your point about the *Posix standard* itself not actually specifying the types, but I see that as more pedantic than practical, since for better or for worse, Pony treats `posix` as a "platform" (the intersection of a collection of platforms), not as a *standard*.  So this is why I see it as appropriate that the `foreigntypes/posix` package represent the intersection of types defined in the other specific platforms that Pony treats as being part of the `posix` meta-platform.

This solution is appealing to me because it is practical for programmers (when inside a `ifdef posix` block, use the `foreigntypes/posix` package, and so on), and it requires no new compiler features.

dipinhora commented:
@jemc I missed that part completely. I think I need to work on my reading comprehension. 8*/

I see a few concerns with this strategy (there's some overlap between the points):

* It would mean that if Pony ever adds a new platform with a unique definition of a standard posix type like `Pid`, then that type would have to be removed it from the `foreigntypes/posix` package (because it's not the same across all platforms anymore) and added to all the platform packages and that would be a breaking change. This same issue could occur if there is a mistake when trying to figure out the platform specific types for each standard posix type across multiple platforms to identify ones that have the same definition.
* New Pony programmers may not understand the distinction between the `posix meta-platform` and the `posix standard` and expect that `foreigntypes/posix` includes all types from the `posix standard`.
* It would mean that programmers would have to remember that some standard posix types are defined in the `foreigntypes/posix` package while others aren't. Programmers may be writing code within an `ifdef posix` block but using `foreigntypes/posix` may not be sufficient because they need a type that is platform specific (i.e. where `linux` defines it differently from `osx`) even though the actual posix function call is the same and this would be counter-intuitive.
* It would mean that programmers now need to know the platform specific quirks of each platforms' `posix` implementation so that they can appropriately `ifdef` to ensure they're using the right type for each specific platforms' version of the standard `posix` type.
* It would mean that instead of being able to do (NOTE: I'm using `off_t` as an example but I don't believe this type is actually different between `linux` and `osx`):
  ```Pony
  use posix = foreigntypes/posix
  let x: posix.off_t = 0
  x = lseek[posix.off_t](_fd, 0, 0)
  ```
  with Pony using the appropriate type for a specific target, a programmer would have to do (where `I64` is a type that can hold all variants):
  ```Pony
  use linux = foreigntypes/linux
  use osx = foreigntypes/osx
  let x: I64 = 0
  x = ifdef linux then
    lseek[linux.off_t](_fd, 0, 0).i64()
  elseifdef osx then
    lseek[osx.off_t](_fd, 0, 0).i64()
  else
  //...
  end
  ```
  which would make writing cross-platform programs that rely on the Posix standard via FFI very cumbersome and error prone.

jemc commented:
@dipinhora - I think your concerns are valid, but I don't think there is a way of resolving your concerns without breaking @sylvanc's concern about types not having a platform-dependent structure.

I'd also say that I think many of your concerns also apply to the implementation you've put forward here in the PR - the only way of solving them would be integrating `ifdefs` into the type definitions as @sylvanc is reluctant to support.

I think my proposal resolves as many concerns as possible without introducing any real new concerns, other than the possible confusion you mentioned about `posix` as a "meta-platform" vs Posix as a *standard*, though this confusion is arguably already somewhat present in Pony as it currently exists.

jemc commented:
Just brainstorming, maybe the Posix confusion could be reduced by renaming `foreigntypes` to `platformtypes`.  Just thinking that `platformtypes/posix` makes it a bit clearer that we're referring to it as a meta-platform and not as a *standard*.

dipinhora commented:
@jemc I agree that the implementation in this PR is not ideal and has the same issues (and likely some additional ones) as your suggestion.

I'm okay with whatever decision you, @sylvanc, and others come up with. I was just raising the concerns since I had assumed the goal was to decide on a long term solution.

Was this discussed during the sync? If yes, was there a decision made?

jemc commented:
We didn't reach this PR during sync - we got caught up on discussions for some of the other ones.  I think we'll probably have time to discuss this week, I hope.

jemc commented:
Discussed toward the end of today's sync call.

We (@sylvanc and I) were unable to come up with a better solution than the subpackage one I proposed, that didn't require language changes to implement.  Specifically, we'd either need to have either 1) type definitions that varied on different platforms, which @sylvanc already expressed heavy reservations about or 2) `libclang` features to read from C headers at Pony source compile time, which is a feature that is being discussed and is awaiting RFC.

If this package is something we want in the short term (before either one of those changes comes about), we'd want to do something like my proposal.

We talked a bit more specifically about the `posix` concerns you raised, where we may have some types that aren't defined in the Posix standard, but just so happen to be the same across all posix platforms that we support.  We agree that it's a bit weird to include them in the `posix` subpackage, but from a practical standpoint it's better than the alternative of not including it, because it makes it clear (through compiler errors) which types can safely be used from an `ifdef posix`, and which types breaking into smaller clauses as you showed above in your `lseek` example.

dipinhora commented:
@jemc Thanks for the update. Let me know whatever you guys decide in terms of whether you want this package (based on your proposal) in the short term or if it should be shelved until the other Pony compiler changes occur.

SeanTAllen commented:
Via sync: of potential interest:

https://doc.rust-lang.org/1.0.0/libc/

SeanTAllen commented:
Im going to leave "needs discussion during sync" on this until we come to some kind of resolution.

sylvanc commented:
I think we don't yet have a solution for Posix types that everybody is happy with. Since there hasn't been a lot of demand, perhaps this is something we don't want to put in the stdlib yet? Maybe @dipinhora 's approach could be a separate package (even if in the ponylang guthub)?

SeanTAllen commented:
Closing this for now for inclusion in standard lib. However we are quite receptive to someone creating a 3rd party library to handle Posix types and cross platform. Perhaps after existing in a 3rd party library for a while, we can reach consensus on the correct "standard library" solution.
